File: java/src/EvalHarness.java
/**
 * Eval Harness — Java
 * Measures retrieval quality: Precision@K, Recall@K, MRR, NDCG@K.
 *
 * java -cp target/rag-ai-java-1.0.0.jar EvalHarness
 */

import com.fasterxml.jackson.databind.ObjectMapper;
import java.util.*;

public class EvalHarness {

    static final ObjectMapper JSON = new ObjectMapper();

    // ── Types ──────────────────────────────────────────────────────────────────
    public record EvalCase(String query, Set<String> relevantIds, String description) {}

    public record EvalResult(
        String query,
        double precisionAtK,
        double recallAtK,
        double reciprocalRank,
        double ndcgAtK,
        List<String> retrievedIds
    ) {}

    public record EvalReport(
        int topK,
        int numCases,
        double meanPrecision,
        double meanRecall,
        double meanMrr,
        double meanNdcg,
        List<EvalResult> perCase
    ) {}

    // ── Metrics ────────────────────────────────────────────────────────────────
    static double precisionAtK(List<String> retrieved, Set<String> relevant, int k) {
        if (k == 0) return 0;
        long hits = retrieved.stream().limit(k).filter(relevant::contains).count();
        return (double) hits / k;
    }

    static double recallAtK(List<String> retrieved, Set<String> relevant, int k) {
        if (relevant.isEmpty()) return 0;
        long hits = retrieved.stream().limit(k).filter(relevant::contains).count();
        return (double) hits / relevant.size();
    }

    static double reciprocalRank(List<String> retrieved, Set<String> relevant) {
        for (int i = 0; i < retrieved.size(); i++) {
            if (relevant.contains(retrieved.get(i))) return 1.0 / (i + 1);
        }
        return 0;
    }

    static double ndcgAtK(List<String> retrieved, Set<String> relevant, int k) {
        double dcg = 0;
        List<String> top = retrieved.subList(0, Math.min(k, retrieved.size()));
        for (int i = 0; i < top.size(); i++) {
            if (relevant.contains(top.get(i))) dcg += 1.0 / (Math.log(i + 2) / Math.log(2));
        }
        // Ideal DCG: all relevant docs at top
        double idcg = 0;
        int idealHits = Math.min(relevant.size(), k);
        for (int i = 0; i < idealHits; i++) {
            idcg += 1.0 / (Math.log(i + 2) / Math.log(2));
        }
        return idcg == 0 ? 0 : dcg / idcg;
    }

    // ── Evaluator ──────────────────────────────────────────────────────────────
    public static class RAGEvaluator {
        final RagEngine engine;
        final int topK;

        RAGEvaluator(RagEngine engine, int topK) {
            this.engine = engine;
            this.topK   = topK;
        }

        EvalResult evaluateCase(EvalCase c) throws Exception {
            List<RagEngine.RetrievedChunk> chunks = engine.retrieve(c.query(), topK);
            List<String> ids = chunks.stream().map(ch -> ch.docId()).toList();
            return new EvalResult(
                c.query(),
                precisionAtK(ids, c.relevantIds(), topK),
                recallAtK(ids, c.relevantIds(), topK),
                reciprocalRank(ids, c.relevantIds()),
                ndcgAtK(ids, c.relevantIds(), topK),
                ids
            );
        }

        EvalReport evaluate(List<EvalCase> cases) throws Exception {
            List<EvalResult> results = new ArrayList<>();
            for (EvalCase c : cases) results.add(evaluateCase(c));
            int n = results.size();
            if (n == 0) return new EvalReport(topK, 0, 0, 0, 0, 0, results);
            double sumP = 0, sumR = 0, sumM = 0, sumN = 0;
            for (EvalResult r : results) { sumP += r.precisionAtK(); sumR += r.recallAtK(); sumM += r.reciprocalRank(); sumN += r.ndcgAtK(); }
            return new EvalReport(topK, n, sumP/n, sumR/n, sumM/n, sumN/n, results);
        }
    }

    static void printReport(EvalReport report) {
        System.out.println("\n" + "=".repeat(50));
        System.out.printf("RAG Eval (Java)  top_k=%d  n=%d%n", report.topK(), report.numCases());
        System.out.println("=".repeat(50));
        System.out.printf("  Mean Precision@%d: %.4f%n", report.topK(), report.meanPrecision());
        System.out.printf("  Mean Recall@%d:    %.4f%n", report.topK(), report.meanRecall());
        System.out.printf("  Mean MRR:          %.4f%n", report.meanMrr());
        System.out.printf("  Mean NDCG@%d:      %.4f%n", report.topK(), report.meanNdcg());
        System.out.println();
        for (EvalResult r : report.perCase()) {
            String status = r.precisionAtK() > 0 ? "✅" : "❌";
            System.out.printf("  %s [P=%.2f R=%.2f MRR=%.2f]  %s%n",
                status, r.precisionAtK(), r.recallAtK(), r.reciprocalRank(),
                r.query().substring(0, Math.min(60, r.query().length())));
        }
    }

    // ── Demo data ──────────────────────────────────────────────────────────────
    static List<RagEngine.Document> demoDocs() {
        return Arrays.asList(
            new RagEngine.Document("Claude is an AI assistant made by Anthropic.", Map.of("source", "anthropic")),
            new RagEngine.Document("RAG combines retrieval with generation to ground LLM answers.", Map.of("source", "rag_paper")),
            new RagEngine.Document("Qdrant is a high-performance vector database written in Rust.", Map.of("source", "qdrant")),
            new RagEngine.Document("Voyage AI provides state-of-the-art embedding and reranking models.", Map.of("source", "voyage")),
            new RagEngine.Document("Java is a class-based, object-oriented language designed for portability.", Map.of("source", "java_docs"))
        );
    }

    // ── CLI ────────────────────────────────────────────────────────────────────
    public static void main(String[] args) throws Exception {
        int topK = 5;
        RagEngine engine = new RagEngine("eval_java");
        List<RagEngine.Document> docs = demoDocs();
        engine.ingest(docs);

        List<EvalCase> cases = new ArrayList<>();
        for (RagEngine.Document doc : docs) {
            String source = doc.metadata.getOrDefault("source", "unknown");
            cases.add(new EvalCase("Tell me about " + source, Set.of(doc.docId), source));
        }

        RAGEvaluator evaluator = new RAGEvaluator(engine, topK);
        EvalReport report = evaluator.evaluate(cases);
        printReport(report);
        System.out.println(JSON.writerWithDefaultPrettyPrinter().writeValueAsString(report));
    }
}

File: go/src/eval_harness.go
// Eval Harness — Go
// Measures retrieval quality: Precision@K, Recall@K, MRR, NDCG@K.
//
// go run ./src/... -eval

package main

import (
	"context"
	"encoding/json"
	"fmt"
	"math"
	"os"
)

// ── Types ──────────────────────────────────────────────────────────────────────
type EvalCase struct {
	Query       string
	RelevantIDs map[string]bool
	Description string
}

type EvalResult struct {
	Query          string
	PrecisionAtK   float64
	RecallAtK      float64
	ReciprocalRank float64
	NDCGAtK        float64
	RetrievedIDs   []string
}

type EvalReport struct {
	TopK          int
	NumCases      int
	MeanPrecision float64
	MeanRecall    float64
	MeanMRR       float64
	MeanNDCG      float64
	PerCase       []EvalResult
}

// ── Metrics ────────────────────────────────────────────────────────────────────
func precisionAtK(retrieved []string, relevant map[string]bool, k int) float64 {
	if k == 0 {
		return 0
	}
	hits := 0
	for _, id := range retrieved[:min(k, len(retrieved))] {
		if relevant[id] {
			hits++
		}
	}
	return float64(hits) / float64(k)
}

func recallAtK(retrieved []string, relevant map[string]bool, k int) float64 {
	if len(relevant) == 0 {
		return 0
	}
	hits := 0
	for _, id := range retrieved[:min(k, len(retrieved))] {
		if relevant[id] {
			hits++
		}
	}
	return float64(hits) / float64(len(relevant))
}

func reciprocalRank(retrieved []string, relevant map[string]bool) float64 {
	for i, id := range retrieved {
		if relevant[id] {
			return 1.0 / float64(i+1)
		}
	}
	return 0
}

func ndcgAtK(retrieved []string, relevant map[string]bool, k int) float64 {
	dcg := func(ids []string) float64 {
		score := 0.0
		for i, id := range ids[:min(k, len(ids))] {
			if relevant[id] {
				score += 1.0 / math.Log2(float64(i+2))
			}
		}
		return score
	}
	actual := dcg(retrieved)
	// Build ideal: relevant docs first
	ideal := make([]string, 0, k)
	for id := range relevant {
		ideal = append(ideal, id)
		if len(ideal) >= k {
			break
		}
	}
	idealDCG := dcg(ideal)
	if idealDCG == 0 {
		return 0
	}
	return actual / idealDCG
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

// ── Evaluator ──────────────────────────────────────────────────────────────────
type RAGEvaluator struct {
	engine *RAGEngine
	topK   int
}

func NewRAGEvaluator(engine *RAGEngine, topK int) *RAGEvaluator {
	return &RAGEvaluator{engine: engine, topK: topK}
}

func (e *RAGEvaluator) EvaluateCase(ctx context.Context, c EvalCase) EvalResult {
	chunks, _ := e.engine.Retrieve(ctx, c.Query, e.topK)
	ids := make([]string, len(chunks))
	for i, ch := range chunks {
		ids[i] = ch.DocID
	}
	return EvalResult{
		Query:          c.Query,
		PrecisionAtK:   precisionAtK(ids, c.RelevantIDs, e.topK),
		RecallAtK:      recallAtK(ids, c.RelevantIDs, e.topK),
		ReciprocalRank: reciprocalRank(ids, c.RelevantIDs),
		NDCGAtK:        ndcgAtK(ids, c.RelevantIDs, e.topK),
		RetrievedIDs:   ids,
	}
}

func (e *RAGEvaluator) Evaluate(ctx context.Context, cases []EvalCase) EvalReport {
	results := make([]EvalResult, len(cases))
	for i, c := range cases {
		results[i] = e.EvaluateCase(ctx, c)
	}
	n := float64(len(results))
	if n == 0 {
		return EvalReport{TopK: e.topK}
	}
	var sumP, sumR, sumMRR, sumNDCG float64
	for _, r := range results {
		sumP   += r.PrecisionAtK
		sumR   += r.RecallAtK
		sumMRR += r.ReciprocalRank
		sumNDCG += r.NDCGAtK
	}
	return EvalReport{
		TopK:          e.topK,
		NumCases:      len(results),
		MeanPrecision: sumP / n,
		MeanRecall:    sumR / n,
		MeanMRR:       sumMRR / n,
		MeanNDCG:      sumNDCG / n,
		PerCase:       results,
	}
}

func (r *EvalReport) Print() {
	fmt.Println("\n" + "="+string(rune('=')) + "===========================================")
	fmt.Printf("RAG Eval Report (Go)  top_k=%d  n=%d\n", r.TopK, r.NumCases)
	fmt.Println("==============================================")
	fmt.Printf("  Mean Precision@%d: %.4f\n", r.TopK, r.MeanPrecision)
	fmt.Printf("  Mean Recall@%d:    %.4f\n", r.TopK, r.MeanRecall)
	fmt.Printf("  Mean MRR:          %.4f\n", r.MeanMRR)
	fmt.Printf("  Mean NDCG@%d:      %.4f\n", r.TopK, r.MeanNDCG)
	fmt.Println()
	for _, res := range r.PerCase {
		status := "✅"
		if res.PrecisionAtK == 0 {
			status = "❌"
		}
		fmt.Printf("  %s [P=%.2f R=%.2f MRR=%.2f]  %s\n",
			status, res.PrecisionAtK, res.RecallAtK, res.ReciprocalRank, truncateStr(res.Query, 60))
	}
	fmt.Println()
}

// ── Demo data ──────────────────────────────────────────────────────────────────
var evalDemoDocs = []*Document{
	{Content: "Claude is an AI assistant made by Anthropic.", Metadata: map[string]string{"source": "anthropic"}},
	{Content: "RAG combines retrieval with generation to ground LLM answers in real documents.", Metadata: map[string]string{"source": "rag_paper"}},
	{Content: "Qdrant is a high-performance vector database written in Rust.", Metadata: map[string]string{"source": "qdrant"}},
	{Content: "Voyage AI provides state-of-the-art embedding and reranking models.", Metadata: map[string]string{"source": "voyage"}},
	{Content: "Go is a statically typed compiled language designed at Google for simplicity and concurrency.", Metadata: map[string]string{"source": "go_docs"}},
}

func runEval(ctx context.Context, topK int) {
	engine, err := NewRAGEngine("eval_go")
	if err != nil {
		fmt.Fprintf(os.Stderr, "eval: engine init failed: %v\n", err)
		os.Exit(1)
	}

	_, err = engine.Ingest(ctx, evalDemoDocs)
	if err != nil {
		fmt.Fprintf(os.Stderr, "eval: ingest failed: %v\n", err)
		os.Exit(1)
	}

	cases := make([]EvalCase, len(evalDemoDocs))
	for i, doc := range evalDemoDocs {
		cases[i] = EvalCase{
			Query:       "Tell me about " + doc.Metadata["source"],
			RelevantIDs: map[string]bool{doc.DocID: true},
			Description: doc.Metadata["source"],
		}
	}

	evaluator := NewRAGEvaluator(engine, topK)
	report := evaluator.Evaluate(ctx, cases)
	report.Print()

	out, _ := json.MarshalIndent(report, "", "  ")
	fmt.Println(string(out))
}
